libcephfs: FSCrypt userspace implementation

chrisphoffman opened the pull request:
This PR adds fscrypt support to CephFS userspace.

Development was done in two phases. This PR will include work done in both phases.

**Phase 1: Core development**
Code: https://github.com/ceph/ceph/pull/52172
Functionality added:
- Implementation of libraries that exist in kernel added (crypto, etc)
- Sync and async IO added
- Links and truncate cases
- Ioctl workarounds added


**Phase 2: Making true to fscrypt spec and feature hardening**
Functionality added:
- Access Semantics
- Multiple Tests
- Additional support for more fs ops
- RMW workload edge cases worked out
- Truncate issues

**Some areas reviewers should focus on:**
- Read/Write paths
- Listing encrypted directories
- Access Semantics
- Fscrypt last block
- Unit tests

**What’s left:**
- Fix issues with unit tests
- Convert tests to include custom ioctl binaries
- Integration tests for multiple clients
- Teuthology testing
- Documentation

Implement fscrypt in libcephfs and cephfs-fuse tracker: https://tracker.ceph.com/issues/63293


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [x] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

phlogistonjohn commented on src/pybind/mgr/dashboard/constraints.txt:
```diff
@@ -5,3 +5,5 @@ python3-saml~=1.4
 requests~=2.26
 Routes~=2.4
 cheroot~=10.0
+grpcio~=1.46
+grpcio-tools~=1.46
```
Did you mean to make dashboard requirements changes in these commits? The message `[DNM] Added PR to fix fc issue.` is rather cryptic to me.

chrisphoffman replied:
This is related to: https://github.com/ceph/ceph/pull/55913
I had more info in the commit, but each push would retrigger "mention" in that PR. I'll make sure to have this removed before it leaves draft.

phlogistonjohn replied:
BTW if you ever would like to try building for a different distro w/o setting up a full blown VM, I've been trying to get more people looking at and testing out my PR #59841

nizamial09 commented:
sorry for the noise but i accidentally added the `wip-nia-testinglabel which i was intending to add on another. removed it btw

hkadam134 commented:
This PR is under test in [https://tracker.ceph.com/issues/69678](https://tracker.ceph.com/issues/69678).

batrick commented:
@chrisphoffman here's the beginnings of a rebase:

https://github.com/batrick/ceph/tree/pr61137-patrick

There are some FIXMEs (grep `FSCRYPT`) from conflicts but otherwise it went okay. Now on to the next part to fix up path traversal.

chrisphoffman commented:
> @chrisphoffman here's the beginnings of a rebase:
> 
> https://github.com/batrick/ceph/tree/pr61137-patrick
> 
> There are some FIXMEs (grep `FSCRYPT`) from conflicts but otherwise it went okay. Now on to the next part to fix up path traversal.

Thanks @batrick , I'll take a look on progress so far.

batrick commented:
Some notes/questions after the rebase:

- do we have a note in the documentation that the debugging information from the client can leak decrypted names? Operators should accept that or turn off client debugging.
- I've tried to avoid touching anything except what is related to the encrypted name / path walking code.
- I left in some blocks labeled with `FSCYRPT` which may need adjustment. In general, we shouldn't need to further modify the dentry / altenrate_name handling before submitting the op (that was already taken care of by the changes in the charmap code) but I did observe some changes like `req->dentry_drop = CEPH_CAP_FILE_SHARED;` which may need to be kept. Please take a careful look.
- You had made a change to `Inode::caps_file_wanted` to obviate setting the mode bit during open. I **partially** changed that: the `Client::path_walk` code now adds that. You may prefer to do it that way but you need to choose one :).
- I haven't been able to test the code because the `fscrypt.cc` source does not compile (not due to my changes).
- Because I have not been able to test, there may be some unexpected breakage. I suggest carefully checking path walk code. Pay special attention to symlink handling in path walks. Probably you should ensure you have test coverage for every branch there.
- There were not any significant conflicts in the commits made by you. Most were in Yehuda's commit which I had squashed. You can double-check what I did by comparing via a manual rebase after squashing Yehuda's commit yourself.
-  Suggest also reviewing the diff of `src/client` compared to `main`.

I've pushed my final update to the rebased branch now. If you have problems feel free to ping me. Good luck!

batrick commented:
> I've pushed my final update to the rebased branch now. If you have problems feel free to ping me. Good luck!

Sorry forgot to squash two small fixes. Final update pushed now as a8d69eaf4f62d184393ff19de5b845f807c1ca18.

batrick commented:
> > I've pushed my final update to the rebased branch now. If you have problems feel free to ping me. Good luck!
> 
> Sorry forgot to squash two small fixes. Final update pushed now as [a8d69ea](https://github.com/ceph/ceph/commit/a8d69eaf4f62d184393ff19de5b845f807c1ca18).

```
$ git push origin $(git branch --show-current) --force
Enumerating objects: 494, done.
Counting objects: 100% (494/494), done.
Delta compression using up to 32 threads
Compressing objects: 100% (342/342), done.
Writing objects: 100% (432/432), 110.89 KiB | 7.92 MiB/s, done.
Total 432 (delta 351), reused 103 (delta 90), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (351/351), completed with 58 local objects.
To github.com:batrick/ceph.git
 + a8d69eaf4f62...63ad28a8aefe pr61137-patrick -> pr61137-patrick (forced update)
```

batrick commented on src/client/Client.cc:
```diff
@@ -15360,18 +16128,24 @@ int Client::_rename(Inode *fromdir, const char *fromname, Inode *todir, const ch
 
-  Dentry *oldde = get_or_create(fromdir, fromname);
-  Dentry *de = get_or_create(todir, toname);
+  req->set_alternate_name(std::move(alternate_name));
```
Was this missing

```suggestion
  req->set_alternate_name(std::move(alternate_name));
  todir->gen_inherited_fscrypt_auth(&req->fscrypt_auth);
```

? I've fixed this in my branch.

chrisphoffman replied:
Yep missed here too

batrick commented on src/client/Client.cc:
```diff
@@ -15490,17 +16264,26 @@ int Client::_link(Inode *in, Inode *dir, const char *newname, const UserPerm& pe
+    req->set_alternate_name(std::move(alternate_name));
+  } else {
+    req->set_alternate_name(de->alternate_name);
+  }
```
ditto

batrick commented on src/client/Client.cc:
```diff
@@ -14728,9 +15407,13 @@ int Client::_mknod(Inode *dir, const char *name, mode_t mode, dev_t rdev,
+  if (r < 0) {
+    delete req;
+    return r;
+  }
```
ditto

batrick replied:
So the file name is not encrypted? I'm a little puzzled by why.

chrisphoffman replied:
At first glance, that's how I interpret it. The wording is not clear on name and/or contents. Maybe it's only referring to contents (which is N/A in this case)?

chrisphoffman replied:
This is done.

batrick commented on src/client/fuse_ll.cc:
```diff
@@ -951,6 +956,184 @@ static void fuse_ll_ioctl(fuse_req_t req, fuse_ino_t ino,
     break;
+    case FS_IOC_GET_ENCRYPTION_POLICY_EX_RESTRICTED:
+    case FS_IOC_GET_ENCRYPTION_POLICY_EX: {
+      auto arg = (fscrypt_get_policy_ex_arg *)in_buf;
```
```
cephadm/build.py: Constructing the zipapp file
cephadm/build.py: Zipapp created with compression
[72/440] Building CXX object src/CMakeFiles/ceph-fuse.dir/client/fuse_ll.cc.o
/home/pdonnell/ceph/src/client/fuse_ll.cc: In function ‘void fuse_ll_ioctl(fuse_req_t, fuse_ino_t, int, void*, fuse_file_info*, unsigned int, const void*, size_t, size_t)’:
/home/pdonnell/ceph/src/client/fuse_ll.cc:961:12: warning: unused variable ‘arg’ [-Wunused-variable]
  961 |       auto arg = (fscrypt_get_policy_ex_arg *)in_buf;
      |            ^~~
/home/pdonnell/ceph/src/client/fuse_ll.cc:1118:11: warning: unused variable ‘r’ [-Wunused-variable]
 1118 |       int r = cfuse->client->get_fscrypt_key_status(arg);
      |           ^
```

batrick commented:
> > > I've pushed my final update to the rebased branch now. If you have problems feel free to ping me. Good luck!
> > 
> > 
> > Sorry forgot to squash two small fixes. Final update pushed now as [a8d69ea](https://github.com/ceph/ceph/commit/a8d69eaf4f62d184393ff19de5b845f807c1ca18).
> 
> ```
> $ git push origin $(git branch --show-current) --force
> Enumerating objects: 494, done.
> Counting objects: 100% (494/494), done.
> Delta compression using up to 32 threads
> Compressing objects: 100% (342/342), done.
> Writing objects: 100% (432/432), 110.89 KiB | 7.92 MiB/s, done.
> Total 432 (delta 351), reused 103 (delta 90), pack-reused 0 (from 0)
> remote: Resolving deltas: 100% (351/351), completed with 58 local objects.
> To github.com:batrick/ceph.git
>  + a8d69eaf4f62...63ad28a8aefe pr61137-patrick -> pr61137-patrick (forced update)
> ```

$ git push origin $(git branch --show-current) --force
Enumerating objects: 1222, done.
Counting objects: 100% (933/933), done.
Delta compression using up to 32 threads
Compressing objects: 100% (425/425), done.
Writing objects: 100% (664/664), 175.77 KiB | 6.51 MiB/s, done.
Total 664 (delta 544), reused 280 (delta 239), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (544/544), completed with 106 local objects.
To github.com:batrick/ceph.git
 + a77830b1867b...07fe04a34692 pr61137-patrick -> pr61137-patrick (forced update)

chrisphoffman commented:
> Some notes/questions after the rebase:
> 
> * do we have a note in the documentation that the debugging information from the client can leak decrypted names? Operators should accept that or turn off client debugging.

This probably shouldn't be included at all. It's been good for dev/debugging though. Will make sure to not have unencrypted names leaking.

> * I left in some blocks labeled with `FSCYRPT` which may need adjustment. In general, we shouldn't need to further modify the dentry / altenrate_name handling before submitting the op (that was already taken care of by the changes in the charmap code) but I did observe some changes like `req->dentry_drop = CEPH_CAP_FILE_SHARED;` which may need to be kept. Please take a careful look.

Will do.

> * You had made a change to `Inode::caps_file_wanted` to obviate setting the mode bit during open. I **partially** changed that: the `Client::path_walk` code now adds that. You may prefer to do it that way but you need to choose one :).

Which lines did you mean?

> * I haven't been able to test the code because the `fscrypt.cc` source does not compile (not due to my changes).

Addressed offline.

batrick commented on doc/cephfs/fs-volumes.rst:
```diff
@@ -585,6 +589,33 @@ Use a command of the following form to remove the earmark of a subvolume:
+.. prompt:: bash #
+
+   ceph fs subvolume enctag get <vol_name> <subvol_name> [--group_name <subvol_group_name>]
+
```
This needs more explanation for what an enctag is, how it will be used in practice, etc.

One might also wonder, why not use `ceph fs subvolume metadata set`?

chrisphoffman replied:
>This needs more explanation for what an enctag is, how it will be used in practice, etc.
Yes I'll add more info about it.

>One might also wonder, why not use ceph fs subvolume metadata set
I considered that and chose this way. Did not feel strongly about either way, but slightly this way.

batrick commented on qa/suites/fs/fscrypt/tasks/0-client.yaml:
```diff
@@ -1,7 +0,0 @@
-    # or the ceph-fuse have supported the fscrypt feature we can remove this
-    # restriction.
-    - if not is_kupstream() then reject() end
-task:
```
Obviously it's thrilling that we can now remove this but we also need a test added that shows a kclient can create an encrypted directory tree and then a ceph-fuse client can read it. Then vice versa.

chrisphoffman replied:
Yes you are correct.
`tasks.cephfs.test_fscrypt.TestFSCryptRMW` has tests with multiple mounts/clients. These tests currently only support the same type of client (fuse and fuse or kclient and kclient). Need to work out how to test multiple types in same test.

chrisphoffman replied:
This is completed, checkout commit: https://github.com/ceph/ceph/pull/61137/commits/e1e2d480fa47f4de2fb96dc25e7330351dfa9a97

batrick commented on qa/tasks/cephfs/test_fscrypt.py:
```diff
@@ -90,6 +91,273 @@ def verify_alternate_name():
+    CLIENTS_REQUIRED = 2
+    def setUp(self):
+        super().setUp()
+        self.mount_b.run_shell_payload(f"sudo fscrypt unlock --quiet --key=/tmp/key {self.path}")
```
so these ioctls work now?

chrisphoffman replied:
Yep we are using a custom fscrypt cli now.

batrick commented on src/common/options/mds-client.yaml.in:
```diff
@@ -589,3 +589,13 @@ options:
+  type: bool
+  level: advanced
+  desc: Enable fscrypt access semantics
+  long_desc: Enable fscrypt access semantics for locked/unlocked states.
```
What does this do exactly? I'm dubious about a client config to switch this on.

gregsfortytwo replied:
I believe this is the option that lets a client look at the encrypted data (rather than returning standard fscrypt access errors). This is necessary to enable cloning of encrypted sub volumes.

batrick commented on src/client/FSCrypt.cc:
```diff
@@ -0,0 +1,1137 @@
+
+void FSCryptContext::generate_new_nonce()
+{
+  cct->random()->get_bytes((char *)nonce, sizeof(nonce));
```
double check this is using a good CPRNG

chrisphoffman replied:
It should be sufficient.

If `getentropy` is available that is used. If not available it is based off of `BCryptGenRandom` for Windows and then `/dev/urandom` for all else. `random()` is used in other places such as rgw, cephx etc.

batrick commented on src/client/Client.cc:
```diff
@@ -6251,6 +6328,9 @@ int Client::may_open(const InodeRef& in, int flags, const UserPerm& perms)
   ldout(cct, 20) << __func__ << " " << *in << "; " << perms << dendl;
   unsigned want = 0;
 
+  if (!in->is_dir() && is_inode_locked(in) && cct->_conf.get_val<bool>("client_fscrypt_as"))
```
cache client_fscrypt_as config

batrick commented on src/client/Client.cc:
```diff
@@ -7742,8 +7829,15 @@ int Client::path_walk(InodeRef dirinode, const filepath& origpath, walk_dentry_r
-    if (i == (path.depth() - 1))
+    if (i == (path.depth() - 1)) {
       caps |= extra_options.mask;
+      if (diri->is_fscrypt_enabled()) {
```
ifdef this out until we figure out what caused the test failure

chrisphoffman replied:
I believe this is working now, resolved.

batrick commented on src/client/Client.cc:
```diff
@@ -8288,18 +8425,124 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
+    //
+    //last block is only needed when truncating smaller
+    //when size is not aligned to FSCRYPT_BLOCK_SIZE
+    //and truncate size is non-zero.
```
in the dev doc on file sizes, have some examples with various situations on truncation with block size not aligned

batrick commented on src/mds/Server.cc:
```diff
@@ -5489,6 +5489,8 @@ void Server::handle_client_setattr(const MDRequestRef& mdr)
       DECODE_START(1, bl);
       decode(header.change_attr, bl);
+      decode(header.file_offset, bl);
+      decode(header.block_size, bl);
```
no verification that we actually got a block of `block_size`. Useful for debugging.

chrisphoffman replied:
Is this a comment that can be resolved or do you request an action from this?

batrick commented on src/client/Client.h:
```diff
@@ -1481,6 +1508,259 @@ class Client : public Dispatcher, public md_config_obs_t {
+
+    void wait() {
+      std::unique_lock l{lock};
+      if (!done) {
```
```suggestion
      while (!done) {
```

chrisphoffman replied:
Done

batrick commented on src/client/Client.cc:
```diff
@@ -11695,6 +12046,277 @@ bool Client::C_Write_Finisher::try_complete()
+    }
+  }
+
+  try_finish(r);
```
`bl` is updated above with the lock but then used in `try_finish`. `bl` is also modified in `::finish_read_start` which could clobber what `finish_read_end` has done prior to calling `try_finish`.

Might not be an actual issue **now** because these callbacks may only be executed from a single `osdc` thread.

chrisphoffman replied:
I suggest this should be part of the refactor task below.

batrick commented on src/client/Client.cc:
```diff
@@ -11695,6 +12046,277 @@ bool Client::C_Write_Finisher::try_complete()
+    }
+  }
+
+  try_finish(r);
```
I remain confused about why we seem to call `::try_finish` from three different code paths (here and in `finish_read_start` and `finish_read_end`).

batrick commented on src/client/Client.cc:
```diff
@@ -11695,6 +12046,277 @@ bool Client::C_Write_Finisher::try_complete()
+  }
+
+  clnt->put_cap_ref(in, CEPH_CAP_FILE_RD);
+  in->mark_caps_dirty(CEPH_CAP_FILE_RD);
```
```suggestion
```

?

batrick commented on src/client/Client.cc:
```diff
@@ -11695,6 +12046,277 @@ bool Client::C_Write_Finisher::try_complete()
+  if (!in->oset.dirty_or_tx)
+    clnt->get_cap_ref(in, CEPH_CAP_FILE_CACHE | CEPH_CAP_FILE_BUFFER);
+
+  clnt->get_cap_ref(in, CEPH_CAP_FILE_BUFFER);
```
Where is

`in->mark_caps_dirty(CEPH_CAP_FILE_WR);`

batrick replied:
looks like in `_write_success`

batrick commented on src/client/Client.cc:
```diff
@@ -11570,10 +11916,15 @@ int64_t Client::_write_success(Fh *f, utime_t start, uint64_t fpos,
+  if (request_size + request_offset > in->effective_size()) {
+    if (encrypted) {
+      in->set_effective_size(request_size + request_offset);
+      in->mark_caps_dirty(CEPH_CAP_FILE_EXCL);
```
This doesn't look right. I think we can only do this if the client is the loner but marking `FILE_WR` cap dirty should be enough.

batrick commented on src/client/Client.cc:
```diff
@@ -11695,6 +12046,277 @@ bool Client::C_Write_Finisher::try_complete()
+
+int Client::WriteEncMgr_NotBuffered::do_write()
+{
+  clnt->get_cap_ref(in, CEPH_CAP_FILE_BUFFER);
```
```suggestion
```

??

batrick commented on src/client/Client.h:
```diff
@@ -1481,6 +1508,259 @@ class Client : public Dispatcher, public md_config_obs_t {
+
+  struct CWF_iofinish;
+
+  class WriteEncMgr : public RefCountedObject {
```
Refactor Task: this is conceptually an overlaid API restriction on RADOS where an object behaves like a block device (where a block size (4k for fscrypt) << max object size (4M cephfs default)): you need to read a block from the device before you can modify it. What I would like to see is that the write path is recast with that abstraction formalized such that we can have the same read-modify-write behavior even for the unencrypted data case (with a config switch set). For the unencrypted data case, we're just reading the data out of RADOS unnecessarily before doing a whole-block write instead of a partial object write. 

The benefit of doing this: you can have a suite of tests which verify the correctness of this abstracted block device for various block sizes without introducing decryption/encryption. Furthermore, the "encrypted data" case would (most probably) literally be an extra if block (decrypt reads) after the reads and an if block (encrypt the write) before the writes.

chrisphoffman replied:
I recommend we do this separately after this PR is merged. We spent a bunch of time resolving issues in this part of the code. I'd like to be able to spend proper attention and time to get a refactor right.

batrick reviewed (approved):
Leaving a blanket approval despite my previous review comments. Most are not blockers or just minor things to check on.

vshankar commented:
Just FYI, there are compilation warnings that I'm seeing in this branch

```
/home/vshankar/ceph/src/client/Client.cc: In member function ‘virtual void Client::C_Read_Sync_NonBlocking::finish(int)’:
/home/vshankar/ceph/src/client/Client.cc:11312:18: warning: comparison of integer expressions of different signedness: ‘uint64_t’ {aka ‘long unsigned int’} and ‘int’ [-Wsign-compare]
11312 |     if (read_len == read) {
      |         ~~~~~~~~~^~~~~~~
/home/vshankar/ceph/src/client/Client.cc: In member function ‘int Client::WriteEncMgr::read_modify_write(Context*)’:
/home/vshankar/ceph/src/client/Client.cc:12163:40: warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]
12163 |   need_read_start = ofs_in_start_block >= 0;
```

dparmar18 commented on src/client/Client.cc:
```diff
@@ -11695,6 +12084,277 @@ bool Client::C_Write_Finisher::try_complete()
+  end_block_ofs = fscrypt_block_start(endoff - 1);
+  ofs_in_end_block = fscrypt_ofs_in_block(endoff - 1);
+
+  need_read_start = ofs_in_start_block >= 0;
```
`ofs_in_start_block` is uint64_t, so this condition is bound to be always true?

chrisphoffman replied:
This has been addressed in: https://github.com/ceph/ceph/commit/e34523ec4bd4f1137ec3433486edb713812d9066

vshankar commented on src/include/cephfs/ceph_ll_client.h:
```diff
@@ -71,6 +71,7 @@ struct ceph_statx {
 	struct timespec	stx_mtime;
 	struct timespec	stx_btime;
 	uint64_t	stx_version;
+	uint64_t 	stx_attributes;
```
This is changing the ABI. Although the field has been added to the end of the structure and since this structure is used to read information, this change is probably fine, but I still wanted to flag this.

synarete replied:
@vshankar I suspect this ABI change could lead to unexpected behaviors: if a user of libcephfs (e.g., Samba's `vfs_ceph_new`) is compiled against the old version, but then linked at run-time with the new version, calling APIs like `ceph_ll_getattr` may cause memory corruption on the caller side (libcephfs assigns extra 8-bytes into provided `stx`). 

As a side note: the kernel's `struct statx` has few [reserved](https://elixir.bootlin.com/linux/v6.14.5/source/tools/include/uapi/linux/stat.h#L137) fields for future extensions. Maybe libcephfs should do the same?

chrisphoffman replied:
Very good points by all.

This was added to support querying if inode was encrypted or not. While we like to stay feature parity with original fscrypt spec, some cases (like this) make sense not to. It seems this capability will cause more harm than good.

Pushed a commit removing this addition/change.

phlogistonjohn commented on src/include/cephfs/libcephfs.h:
```diff
@@ -131,6 +131,10 @@ struct ceph_ll_io_info {
   bool syncdataonly;
 };
 
+struct ceph_fscrypt_key_identifier;
+struct fscrypt_policy_v2;
+struct fscrypt_remove_key_arg;
```
How are users of the C api intended to work with these structs?  For example if I wrote:

```
struct ceph_fscrypt_key_identifier kid;
ret = ceph_add_fscrypt(cmount, key_data, key_len, &kid, user);
```
and then I later wanted to remove that key would I have to convert kid to a `struct fscrypt_remove_key_arg`?  


Similarly, I don't see how to create a valid `struct fscrypt_policy_v2` to pass to ceph_set_fscrypt_policy_v2.

chrisphoffman replied:
I was off for a week, I'm back now. 

>and then I later wanted to remove that key would I have to convert kid to a struct fscrypt_remove_key_arg?

yes that is correct

In order to create the structs you mentioned, check out snippets from `src/test/libcephfs/fscrypt.cc`:

```
void generate_remove_key_arg(ceph_fscrypt_key_identifier kid, fscrypt_remove_key_arg* arg){
  fscrypt_key_specifier key_spec;
  key_spec.type = FSCRYPT_KEY_SPEC_TYPE_IDENTIFIER;
  key_spec.__reserved = 0;
  memcpy(key_spec.u.identifier, kid.raw, 16);
  arg->removal_status_flags = 0;
  arg->key_spec = key_spec;
}  
  
void populate_policy(struct ceph_fscrypt_key_identifier kid, struct fscrypt_policy_v2* policy) {
  policy->version = 2;
  policy->contents_encryption_mode = FSCRYPT_MODE_AES_256_XTS;
  policy->filenames_encryption_mode = FSCRYPT_MODE_AES_256_CTS;
  policy->flags = FSCRYPT_POLICY_FLAGS_PAD_32;
  memcpy(policy->master_key_identifier, kid.raw, FSCRYPT_KEY_IDENTIFIER_SIZE);
}
```

Then within `fscrypt.cc`, you can see how they are used. 

If using larger keys checkout `struct fscrypt_add_key64_arg` defined in `src/client/fscrypt_uapi.h`.

chrisphoffman commented:
Repushed with the following:
-Rebased
-Fixed failure in testing clone: https://pulpito.ceph.com/choffman-2025-05-29_15:03:33-fs:volumes-wip-choffman-fscrypt-528-distro-default-gibba/8301557/ . The failure was caused when source file size was 0.

chrisphoffman commented on src/client/Client.cc:
```diff
@@ -17430,6 +18251,27 @@ mds_rank_t Client::_get_random_up_mds() const
 }
 
+int  Client::get_inode_flags(const Inode* in, int* file_attr_out) {
+  if (!file_attr_out)
```
addressed in: https://github.com/ceph/ceph/pull/61137/commits/ced35cedaac8a25b4d317f3b5fe4b406b6ac363f

phlogistonjohn commented:
How are things going with this PR?  The smb team has been asked to implement things related to this work. We prefer to work in upstream first but hesitate to start on it until this PR is merged. Thanks!

chrisphoffman commented on src/client/Client.cc:
```diff
@@ -11695,6 +12046,277 @@ bool Client::C_Write_Finisher::try_complete()
+
+    /* prepend data from the start of the first block */
+    bufferlist newbl;
+    startbl.splice(0, ofs_in_start_block, &newbl);
```
I'm auditing the write path now. 

This section is dealing with plain text data. The variable ofs_in_start_block in this case is the logical offset (unencrypted form). This will be a no-op when ofs_in_start_block is 0. Otherwise, we'll only want to modify data starting at ofs_in_start_block. That means we will splice/retain data offset 0 to ofs_in_start_block.

Here's a general flow of a write:
-write(4098, 3, "buf")
-enter read_modify_write() 
-read start block via read(4096, 4096)
-_read(..)

-Inside finish_read_start():
populate newbl with startbl data before offset above (unmodified data)
append "buf" to newbl
if necessary, append unmodified data to the end
if necessary, append 0s to the end up to block size boundary

chrisphoffman commented on qa/tasks/cephfs/mount.py:
```diff
@@ -1492,6 +1508,84 @@ def stat(self, fs_path, follow_symlinks=True, wait=True, **kwargs):
 
+    def lchown(self, fs_path, uid, gid):
+        """
+        Change ownership of a link with uid and gid provided.
```
Done.

chrisphoffman commented on qa/tasks/cephfs/mount.py:
```diff
@@ -1515,6 +1609,28 @@ def touch(self, fs_path):
 
+    def touch_os(self, fs_path):
+        """
+        Create a dentry if it doesn't already exist.  Use open in os module.
```
Done.

chrisphoffman commented on qa/tasks/cephfs/mount.py:
```diff
@@ -717,20 +717,30 @@ def check_files(self):
+    def write_file(self, path, data, perms=None, **kwargs):
+        """
+        Write the given data at the given path and set the given perms to the
+        file on the path.
```
I left this as is because it is a copy of what was already in upstream. I'm also unsure of what you meant in your comment.

chrisphoffman commented on src/libcephfs.cc:
```diff
@@ -2477,6 +2477,90 @@ extern "C" void ceph_finish_reclaim(class ceph_mount_info *cmount)
+  return cmount->get_client()->remove_fscrypt_key(kid, user);
+}
+
+extern "C" int ceph_get_fscrypt_key_status(struct ceph_mount_info *cmount,
```
This is done.

chrisphoffman commented on src/test/libcephfs/fscrypt.cc:
```diff
@@ -0,0 +1,1151 @@
+  ceph_close(cmount, fd);
+  ceph_shutdown(cmount);
+}
+
```
We override some of the setup/tear down in `ceph_test_libcephfs_fscrypt` to enable/utilize fscrypt code paths. Many of the tests already run within fscrypt enable dirs. Can we utilize `ManyNestedDirsCaseInsensitive` for this? 

For example:
```
ceph_test_libcephfs_fscrypt --gtest_filter=*ManyNestedDirsCaseInsensitive
init_fscrypt(): fscrypt_dir=ceph_test_libcephfs_fscrypt.1817367.950732402
init_fscrypt(): init fscrypt done
Note: Google Test filter = *ManyNestedDirsCaseInsensitive
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from LibCephFS
[ RUN      ] LibCephFS.ManyNestedDirsCaseInsensitive
[       OK ] LibCephFS.ManyNestedDirsCaseInsensitive (2250 ms)
[----------] 1 test from LibCephFS (2251 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (2251 ms total)
[  PASSED  ] 1 test.

```

chrisphoffman commented on src/client/FSCrypt.h:
```diff
@@ -0,0 +1,407 @@
+
+using FSCryptFDataDencRef = std::shared_ptr<FSCryptFDataDenc>;
+
+class FSCryptDecryptedInodes {
```
This is done, checkout : https://github.com/ceph/ceph/pull/61137/commits/1d94f2679307bd3c73fac1afdd427d3f3be70a77

chrisphoffman commented on src/client/Client.h:
```diff
@@ -1481,6 +1508,259 @@ class Client : public Dispatcher, public md_config_obs_t {
+
+    std::unique_ptr<_Ctx> _ctx;
+
+    T& t;
```
I suggest we do this during the refactor test below.

chrisphoffman commented on src/client/Client.cc:
```diff
@@ -17430,6 +18251,27 @@ mds_rank_t Client::_get_random_up_mds() const
+
+  *file_attr_out = 0;
+    // set or clear the encryption flag depending on the inode status
+  if (in->is_encrypted()) {
```
This function no longer exists so no need to address.

chrisphoffman commented on src/client/fuse_ll.cc:
```diff
@@ -951,6 +956,184 @@ static void fuse_ll_ioctl(fuse_req_t req, fuse_ino_t ino,
+    case FS_IOC_GET_ENCRYPTION_POLICY_EX: {
+      auto arg = (fscrypt_get_policy_ex_arg *)in_buf;
+
+      generic_dout(0) << __FILE__ << ":" << __LINE__ << ": in_bufsz=" << in_bufsz << " out_bufsz=" << out_bufsz << " FS_IOC_GET_ENCRYPTION_POLICY_EX buffer:\n" << fscrypt_hex_str(in_buf, in_bufsz) << dendl;
```
Done.

chrisphoffman commented on src/test/libcephfs/test.cc:
```diff
@@ -3684,7 +3709,29 @@ TEST(LibCephFS, SetMountTimeout) {
+  //should fail with not supported, as vxattr exists but is read only
+  ASSERT_EQ(-EOPNOTSUPP, ceph_fsetxattr(cmount, fd, "ceph.fscrypt.auth", "foo", 3, CEPH_XATTR_CREATE));
+  //should fail with EINVAL as vxattr is not exposed
+  ASSERT_EQ(-EINVAL, ceph_fsetxattr(cmount, fd, "ceph.fscrypt.file", "foo", 3, CEPH_XATTR_CREATE));
```
This is done: https://github.com/ceph/ceph/pull/61137/commits/b61fae9555e2e1540cce38df04f25f2f1bf19130

chrisphoffman commented on src/client/Client.cc:
```diff
@@ -7345,7 +7346,7 @@ int Client::fscrypt_dummy_encryption() {
     key_spec.type = FSCRYPT_KEY_SPEC_TYPE_IDENTIFIER;
     key_spec.__reserved = 0;
-    memcpy(key_spec.u.identifier, kid.raw, 16);
+    memcpy(key_spec.u.identifier, keyid, 16);
```
Done.

chrisphoffman commented on src/test/libcephfs/fscrypt.cc:
```diff
@@ -1125,11 +1104,11 @@ TEST(FSCrypt, SetPolicyAlreadyExistDifferentPolicyNotEmpty) {
   }
-  struct ceph_fscrypt_key_identifier kid2;
-  r = ceph_add_fscrypt_key(cmount, fscrypt_key2, sizeof(fscrypt_key2), &kid2, 1299);
+  char keyid2[16];
```
Done.
